### PR TITLE
fix: Record cudagraphs when weight streaming budget has changed

### LINKED ISSUE #3308: 🐛 [Bug] Error when weigh streaming and cuda graphs is used
##  Bug Description

"RuntimeError: CUDA error: invalid argument" if cuda graphs is enabled and weight streaming budget has changed.
It seems cuda graphs need to record when weight streaming budget is changed

## To Reproduce

model = SampleModel().eval().cuda()
input = [torch.randn(*INPUT_SIZE, dtype=torch.float32).cuda()]
fx_graph = torch.fx.symbolic_trace(model)

optimized_model = torchtrt.compile(
    fx_graph,
    inputs=input,
    ir="dynamo",
    min_block_size=1,
    cache_built_engines=False,
    reuse_cached_engines=False,
    use_python_runtime=True,
    use_explicit_typing=True,
    enable_weight_streaming=True,
)

torchtrt.runtime.set_cudagraphs_mode(True)    

# Weight streaming context keeps current device budget size
with torchtrt.runtime.weight_streaming(optimized_model) as weight_streaming_ctx:
    new_budget = int(weight_streaming_ctx.total_device_budget * 0.2)
    weight_streaming_ctx.device_budget = new_budget
    optimized_model(*input)

    new_budget = int(weight_streaming_ctx.total_device_budget * 0.4)
    weight_streaming_ctx.device_budget = new_budget
    optimized_model(*input)

## Expected behavior

no cuda runtime error

## Environment

> Build information about Torch-TensorRT can be found by turning on debug messages

 - Torch-TensorRT Version (e.g. 1.0.0):
 - PyTorch Version (e.g. 1.0):
 - CPU Architecture:
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, `libtorch`, source):
 - Build command you used (if compiling from source):
 - Are you using local sources or building from archives:
 - Python version:
 - CUDA version:
 - GPU models and configuration:
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by keehyuna**
# Description

Add flag indicates to reevaluate the runtime settings when weight streaming budget is changed.
Fixes #3308

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: core/runtime/execute_engine.cpp [hunk 1]
```diff
@@ -115,11 +115,15 @@ std::vector<at::Tensor> execute_engine(std::vector<at::Tensor> inputs, c10::intr
   }
 
   // Whether cudagraphs needs to record the graph on this pass
-  bool need_cudagraphs_record = (CUDAGRAPHS_MODE && (!_cudagraphs_validate_shapes(inputs, compiled_engine)));
+  bool need_cudagraphs_record =
```
Comment (narendasan) [inline core/runtime/execute_engine.cpp:118]: Seems like we need something a bit more comprehensive than a bunch of booleans for cudagraph rerecord

Comment (keehyuna) [reply core/runtime/execute_engine.cpp:118]: I agree with it. I'm thinking of below structure. Could you review [PR3276]?(https://github.com/pytorch/TensorRT/pull/3276)
https://github.com/pytorch/TensorRT/pull/3276/files#diff-3d6304b0f21f64bdf7867ceac346866d505619401e9c7b6a696cfe3bb567254eR37-R63

**Comment by keehyuna** [comment]
Will adopt proper structure for runtime states after https://github.com/pytorch/TensorRT/pull/3276 is reviewed.

File: tests/py/dynamo/runtime/test_004_weight_streaming.py [hunk 2]
```diff
@@ -238,6 +242,171 @@ def test_weight_streaming_multi_rt(self, _, use_python_runtime):
...
+    @parameterized.expand(
+        [
+            ("python_runtime", True),
+            ("cpp_runtime", False),
+        ]
+    )
+    def test_weight_streaming_cudagraphs(self, _, use_python_runtime):
+        model = SampleModel().eval().cuda()
+        input = [torch.randn(*INPUT_SIZE, dtype=torch.float32).cuda()]
+        fx_graph = torch.fx.symbolic_trace(model)
```
Comment (peri044) [inline tests/py/dynamo/runtime/test_004_weight_streaming.py:254]: can you use `torch.export.export()` to export the model?

File: tests/py/dynamo/runtime/test_004_weight_streaming.py [hunk 3]
```diff
@@ -238,6 +242,171 @@ def test_weight_streaming_multi_rt(self, _, use_python_runtime):
...
+
+        inputs = torchtrt.Input(
+            min_shape=(1, 100),
+            opt_shape=(64, 100),
+            max_shape=(128, 100),
+            dtype=torch.float,
+            name="x",
+        )
+        model = SampleModel().eval().cuda()
+        fx_graph = torch.fx.symbolic_trace(model)
```
Comment (peri044) [inline tests/py/dynamo/runtime/test_004_weight_streaming.py:328]: same here. Use `torch.export.export()`

**Review by peri044 — COMMENTED** [review]
Functionality LGTM

**Pull request merged**